src/lib/openssl_preload/crypto.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Caches whether or not a memory address is from libssl.so
+typedef struct {
+    void* addr;
+    bool is_libssl;
+} bt_cache_entry_t;
+
+// We use a small cache size: in ad-hoc experiments with tor-0.4.6.9, we observed
+// at most three callers of EVP_EncryptUpdate.
+#define EVP_BACKTRACE_CACHE_LEN 10
+static bt_cache_entry_t evp_backtrace_cache[EVP_BACKTRACE_CACHE_LEN];
+
+// For storing a pointer to the original EVP_EncryptUpdate function.
+typedef int EVP_EncryptUpdate_func(void*, unsigned char*, int*, const unsigned char*, int);
+static void* evp_eu_funcptr = NULL;
+
+#ifdef DEBUG
+// Counters for verifying that interception is happening correctly.
+static unsigned long aes_e_cnt = 0;
+static unsigned long aes_d_cnt = 0;
+static unsigned long aes_ce_cnt = 0;
+static unsigned long cry_ce_cnt = 0;
+static unsigned long cry_cec_cnt = 0;
+static unsigned long evp_c_cnt = 0;
+static unsigned long evp_eu_cnt = 0;
+
+static void _print_counters() {
+    fprintf(stderr,
+            "Counters: {'AES_encrypt':%lu, 'AES_decrypt':%lu, 'AES_ctr128_encrypt':%lu, "
+            "'CRYPTO_ctr128_encrypt':%lu, 'CRYPTO_ctr128_encrypt_ctr32':%lu, "
+            "'EVP_Cipher':%lu, 'EVP_EncryptUpdate':%lu}\n",
+            aes_e_cnt, aes_d_cnt, aes_ce_cnt, cry_ce_cnt, cry_cec_cnt, evp_c_cnt, evp_eu_cnt);
+}
+#endif
+
+__attribute__((constructor)) void _crypto_load() {
+#ifdef DEBUG
+    fprintf(stderr, "Loading the preloaded crypto interception lib\n");
+#endif
+
+    // Initialize backtrace address cache pointers to NULL.
+    memset(evp_backtrace_cache, 0, sizeof(bt_cache_entry_t) * EVP_BACKTRACE_CACHE_LEN);
+
+    // Get a ref to the EVP_EncryptUpdate that would be called if we didn't preload.
+    evp_eu_funcptr = dlsym(RTLD_NEXT, "EVP_EncryptUpdate");
+
+#ifdef DEBUG
+    fprintf(stderr, "dlsym for EVP_EncryptUpdate returned %p\n", evp_eu_funcptr);
+#endif
+}
+
+#ifdef DEBUG
+__attribute__((destructor)) void _crypto_unload() {
+    fprintf(stderr, "Unloading the preloaded crypto interception lib\n");
+    _print_counters();
+}
+
+#define _maybe_print_counters(cnt)                                                                 \
+    {                                                                                              \
+        if ((cnt % 1000) == 0) {                                                                   \
+            _print_counters();                                                                     \
+        }                                                                                          \
+    }
+#endif
+
+void AES_encrypt(const unsigned char* in, unsigned char* out, const void* key) {
+#ifdef DEBUG
```

Comment:
Optional - it looks like you could move these `#ifdef`s into or around the macro/function definition.

Reply:
Done.